app/src/main/java/net/squanchy/schedule/tracksfilter/widget/FilterChipView.kt

```diff
@@ -0,0 +1,239 @@
+
+    private fun updateColors(@ColorInt baseColor: Int) {
+        this.baseColor = baseColor
+        backgroundColor = CheckableValue(checkedValue = baseColor, uncheckedValue = Color.TRANSPARENT)
+        strokeColor = CheckableValue(checkedValue = baseColor, uncheckedValue = baseColor)
+
+        val checkedTextColor = baseColor.pickBestTextColorByContrast(checkedTextColorLight, checkedTextColorDark)
+        val contrastSafeBaseColor = baseColor.darkenToEnsureTextContrasts(parentBackgroundColor, uncheckedFallbackTextColor)
+        val uncheckedTextColor = parentBackgroundColor.pickBestTextColorByContrast(baseColor, contrastSafeBaseColor)
+        textColor = CheckableValue(checkedValue = checkedTextColor, uncheckedValue = uncheckedTextColor)
+
+        redraw()
+    }
+
+    override fun isChecked(): Boolean = isChecked
+
+    override fun toggle() {
+        setChecked(!isChecked)
+    }
+
+    override fun setChecked(checked: Boolean) {
+        if (checked == isChecked) return
+        isChecked = checked
+
+        redraw()
+
+        // Avoid infinite recursions if checked is set from a listener
+        if (isBroadcastingCheckedChange) {
+            return
+        }
+
+        isBroadcastingCheckedChange = true
+        onCheckedChangeListener?.invoke(this, checked)
+        isBroadcastingCheckedChange = false
+    }
+
+    private fun redraw() {
+        updateBackgroundDrawable()
+        updateTextColor()
+        invalidate()
+    }
+
+    private fun updateBackgroundDrawable() {
+        with(backgroundDrawable) {
+            setStroke(strokeWidth, strokeColor[isChecked])
+            color = ColorStateList.valueOf(backgroundColor[isChecked])
+            alpha = (backgroundAlpha[isChecked] * Companion.MAX_ALPHA_VALUE).toInt()
+        }
+
+        invalidateOutline()
+    }
+
+    private fun updateTextColor() {
+        super.setTextColor(textColor[isChecked])
+    }
+
+    var onCheckedChangeListener: ((FilterChipView, Boolean) -> Unit)? = null
+
+    override fun getAccessibilityClassName(): CharSequence {
+        return FilterChipView::class.java.name
+    }
+
+    override fun onInitializeAccessibilityNodeInfo(info: AccessibilityNodeInfo) {
+        super.onInitializeAccessibilityNodeInfo(info)
+        info.isCheckable = true
+        info.isChecked = isChecked
+    }
+
+    override fun onInitializeAccessibilityEvent(event: AccessibilityEvent) {
+        super.onInitializeAccessibilityEvent(event)
+        event.isChecked = isChecked
+    }
+
+    override fun setOnClickListener(l: OnClickListener?) {
+        throw UnsupportedOperationException("Can't set a custom click listener on FilterChipView, use onCheckedChangeListener instead")
+    }
+
+    override fun setBackgroundTintList(tint: ColorStateList?) {
+        allowOnlyWhileInitializing("Can't interact with a FilterChipView's background directly") {
```

Comment:
Shouldn't we use super inside the class and always throw here?

Reply:
Unfortunately we can't because `TextView` does a loooot of calls during its initialisation. It originally was like that, it's the `super` itself calling these methods tho.

Reply:
Ah I see, fair enough